src/mono/mono/mini/interp/transform.c

```diff
@@ -2593,6 +2593,40 @@ interp_transform_internal_calls (MonoMethod *method, MonoMethod *target_method,
 	return target_method;
 }
 
+/* Return whenever TYPE represents a vtype with only one scalar member */
+static gboolean
+is_scalar_vtype (MonoType *type)
+{
+	MonoClass *klass;
+	MonoClassField *field;
+	gpointer iter;
+
+	if (!MONO_TYPE_ISSTRUCT (type))
+		return FALSE;
+	klass = mono_class_from_mono_type_internal (type);
+	mono_class_init_internal (klass);
+
+	int size = mono_class_value_size (klass, NULL);
+	if (size == 0 || size >= 8)
+		return FALSE;
+
+	iter = NULL;
+	int nfields = 0;
+	field = NULL;
+	while ((field = mono_class_get_fields_internal (klass, &iter))) {
+		if (field->type->attrs & FIELD_ATTRIBUTE_STATIC)
+			continue;
+		nfields ++;
+		if (nfields > 1)
+			return FALSE;
+		MonoType *t = mini_get_underlying_type (field->type);
+		if (!((MONO_TYPE_IS_PRIMITIVE (t) || MONO_TYPE_IS_REFERENCE (t) || MONO_TYPE_IS_POINTER (t))))
```

Comment:
This seems to incorrectly allow some 8 byte primitives on 32bit targets

Reply:
I'm not sure I understand how my observation is addressed here. Let's say on 32 bit system, you have to pass a struct that contains just a `long` field. This check here will consider it as valid to pass it as a PTR. But in `do_icall` in `interp.c`, we can only pass a 32bit native int by value.

Reply:
There is a size > SIZEOF_VOID_P check at the beginning.